### PR TITLE
[AIRFLOW-4922]Fix task get log by Web UI

### PR DESCRIPTION
**Pull request by ITriangle**
[AIRFLOW-4922]Fix task get log by Web UI

When getting logs, querying task_instance does not get the exact hostname. Need to cooperate with the log table and try number

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4922
  
    In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

  When getting logs, querying task_instance does not get the exact hostname. Need to cooperate with the log table and try number

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

  test my local Airflow Cluster

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by cloneluke** [comment]
I would like this fix also

**Comment by dimberman** [comment]
@ITriangle can you post a before/after of the UI?

File: airflow/utils/log/file_task_handler.py [hunk 1]
```diff
@@ -80,6 +80,35 @@ def _render_filename(self, ti, try_number):
                                              task_id=ti.task_id,
                                              execution_date=ti.execution_date.isoformat(),
                                              try_number=try_number)
+    from airflow.utils.db import provide_session
```
Comment (XD-DENG) [inline airflow/utils/log/file_task_handler.py]: Let's put all `import`s at the beginning part of the file, as suggested in PEP 8 (https://www.python.org/dev/peps/pep-0008/)

Comment (XD-DENG) [reply airflow/utils/log/file_task_handler.py]: The same for your line `from airflow import models` and `import json`

File: airflow/utils/log/file_task_handler.py [hunk 2]
```diff
@@ -80,6 +80,35 @@ def _render_filename(self, ti, try_number):
...
+    @provide_session
+    def query_task_hostname(self, ti, try_number, session=None):
+        """
+        Get task log hostname by log table and  try_number
```
Comment (XD-DENG) [inline airflow/utils/log/file_task_handler.py]: Duplicated spaces between `and` and `try_number`

**Review by XD-DENG — CHANGES_REQUESTED** [review]
Minor changes to make

File: airflow/utils/log/file_task_handler.py [hunk 3]
```diff
@@ -81,6 +82,34 @@ def _render_filename(self, ti, try_number):
...
+        """
+        from airflow import models
+        res_log = session.query(models.Log).filter(
+            models.Log.dag_id == ti.dag_id,
+            models.Log.task_id == ti.task_id,
+            models.Log.execution_date == ti.execution_date,
+            models.Log.owner != 'anonymous',
+            models.Log.event == 'cli_run',
+            models.Log.extra.like('%--raw%')
+        ).order_by(models.Log.id).limit(1).offset(try_number - 1).first()
```
Comment (ashb) [inline airflow/utils/log/file_task_handler.py:104]: Which cases is this needed in? This seems like an awful lot of conditions here.

Comment (ashb) [reply airflow/utils/log/file_task_handler.py:104]: > querying task_instance does not get the exact hostname.

Why not? Do you need to configure the hostname detection in airflow.cfg:

```
# Hostname by providing a path to a callable, which will resolve the hostname
# The format is "package:function". For example,
# default value "socket:getfqdn" means that result from getfqdn() of "socket" package will be used as hostname
# No argument should be required in the function specified.
# If using IP address as hostname is preferred, use value "airflow.utils.net:get_host_ip_address"
hostname_callable = socket:getfqdn
```

**Comment by ashb** [comment]
Right, I just read the jira issue properly.

This is the wrong fix. Instead we should commit the hostname column early.

**Comment by ITriangle** [comment]
This is the current method！If it is better to increase the hostname information. No need to query from the log table.

**Comment by ITriangle** [comment]
> Right, I just read the jira issue properly.
> 
> This is the wrong fix. Instead we should commit the hostname column early.

This is the current method！If it is better to increase the hostname information . No need to query from the log table.

The current way is to get from the task_instance table, but task_instance only saves the latest record, which cannot meet the task of multiple attempts.

Looking forward to new implementation

**Comment by samgans** [comment]
Hi @ITriangle, @ashb.  I've stumbled upon the issue and wanna try to resolve it.

Can you please tell me some more on how the FileTaskHandler is called and why it can't read the tasks URL as it has been committed during the handling of the task's failure?

**Comment by pelaprat** [comment]
Why is this PR closed? This bug still exists in Airflow 2.x.

**Comment by s22-rxr** [comment]
confirming this still exists

**Comment by ashb** [comment]
As I stated earlier -- this is the wrong fix.

There is a hostname column on the TI table which should be correct -- we should ensure that value is correct, rather than joining against an extra table.

**Comment by pelaprat** [comment]
Thanks, @ashb. Is there another issue that properly describes the problem. We are encountering this problem across many of our tasks in Airflow 2.0.2.

**Comment by bensonnd** [comment]
@ashb we use a mounted volume inside the docker container for all the logs, so would host name in that case still be relevant?

**Comment by ashb** [comment]
@bensonnd It wouldn't be relevant _for you_, no.

**Comment by pelaprat** [comment]
@ashb – Is there an issue created for what @bensonnd describes? We're seeing the problem all over the place in our DAG runs, creating a lot of manual work to troubleshoot.

**Comment by tatitati** [comment]
yeah, I confirm it exists yet ( using Airflow 1.10.14 )

**Comment by ashb** [comment]
Not specifically - we don't plan at the bug level, just high level features.

So either someone has to tackle this themselves (which we will help you do), or engage in some kind of commercial relationship with a committer/company employing committers to get this bug fixed in a fixed time frame.

That said I remember a change that was committed a few days ago (for something else, not specifically this bug) that may fix this in 2.1.0 when it's out, which should be "days to weeks" timeframe

**Comment by yerriswamykonanki** [comment]
Hi,
I am still seeing the  issue in 2.0.2 version

**Comment by ashb** [comment]
Yes, 7 days ago I said "this fix which is in 2.1 might fix it.

**Comment by abhishektrivedi-hs** [comment]
Still a bug

*** Failed to fetch log file from worker. Unsupported URL protocol ''

**Comment by peschue** [comment]
> That said I remember a change that was committed a few days ago (for something else, not specifically this bug) that may fix this in 2.1.0 when it's out, which should be "days to weeks" timeframe.
> 
> Edit: the change in taskinstance.py in [817b599](https://github.com/apache/airflow/commit/817b599234dca050438ee04bc6944d32bc032694) is what I was thinking of

The 2.1.0-rc2 was released recently and there is also a "2.1.0-python3.8" tag in dockerhub - does the 2.1.0-python3.8 docker image contain 2.1.0-rc2?

**Comment by potiuk** [comment]
This was merged into main but not yet released.

**Comment by potiuk** [comment]
UPDATE: double-checked - it was released in 2.1.0 (last commit merge there).

**Comment by peschue** [comment]
Thanks a lot, I will try with the 2.1.0-python3.8 docker image if this solves the issue.

**Comment by raajtilaksarma** [comment]
```
*** Log file does not exist: /opt/airflow/logs/my-task/my-task/2021-06-15T14:11:33.254428+00:00/1.log
*** Fetching from: http://:8793/log/my-task/my-task/2021-06-15T14:11:33.254428+00:00/1.log
*** Failed to fetch log file from worker. Unsupported URL protocol ''
```

This still exists after upgrading to `apache/airflow:2.1.0`.
Is this a newer issue or related to the existing one?

**Comment by molejnik-mergebit** [comment]
I'm on 2.1.0 and also experience the issue. Sometimes when task fails there is no way to obtain logs.

**Comment by peschue** [comment]
For me, upgrading to 2.1.0 resolved the issue.

**Comment by peschue** [comment]
The issue just happened again, with 2.1.0

*** Fetching from: http://:8793/log/.../latest_only/2021-06-21T22:00:00+00:00/1.log
*** Failed to fetch log file from worker. Unsupported URL protocol ''

**Comment by fj-sanchez** [comment]
We just updated to 2.1.0 and we have started to see the same error.

```
*** Log file does not exist: /airflow/logs/job_pipeline_trigger/get-params/2021-07-06T14:24:51.780820+00:00/1.log
*** Fetching from: http://:8793/log/job_pipeline_trigger/get-params/2021-07-06T14:24:51.780820+00:00/1.log
*** Failed to fetch log file from worker. The request to ':///' is missing either an 'http://'                         or 'https://' protocol.
```

I will investigate a bit more but any suggestions are welcomed.

**Comment by ashb** [comment]
99% of the time when this happens, it is because the Celery executor has a problem before the airflow code gets a chance to run.

I understand it makes it tricky to diagnose, but this fix is not the right one, (as if as I suspect) there wouldn't actually _be_ a log file written.

If any one is able to validate if this is the case that would be helpful. If you see this and you can check all of your worker nodes to see if the log file exists on _any_ of them. (My hunch is it wont).

**Comment by abhishektrivedi-hs** [comment]
I was facing the same "Log file does not exist:" issue during initial days of Airflow installation. Later once i understood the end to end design and architecture, debugging was easy. Below are the scenarios where i faced this issue and the reason for it--
1> Webserver workers are different than celery or task workers. Due to the confusion, i did not start any airflow worker in my dev server with local executor option. While executing the  dag from UI, it still worked because scheduler can act as a worker if mode is LocalExecutor. But later i was getting log not found error. Inorder to fix this, i just started airflow workers.
2> Same log not found error i got with celery executor. This was due to failure at worker end. This failure was due to parsing or config related. We can verify this under worker.out log.

**Comment by fj-sanchez** [comment]
> If any one is able to validate if this is the case that would be helpful. If you see this and you can check all of your worker nodes to see if the log file exists on _any_ of them. (My hunch is it wont).

I think that you're right, it seems to be a problem before even starting:
```{"event": "Failed to execute task No SecretsMasker found!.", "level": "error", "logger": "airflow.executors.local_executor.QueuedLocalWorker", "timestamp": "2021-07-07T17:44:32.327926"}```

I think that we need to update some of the configurations after upgrading to 2.1.0...

**Comment by vdusek** [comment]
@ashb We're having the same troubles after upgrade to 2.1.X. I'd say around 5% of our tasks are failing with the following:

Log on webserver:
```
*** Falling back to local log
*** Log file does not exist: /home/airflow/logs/dm-bi-edw-c360-export-brg/ext_w_brg_bi_c360_t_customer_support/2021-07-08T10:00:00+00:00/1.log
*** Fetching from: http://:8793/log/dm-bi-edw-c360-export-brg/ext_w_brg_bi_c360_t_customer_support/2021-07-08T10:00:00+00:00/1.log
*** Failed to fetch log file from worker. The request to ':///' is missing either an 'http://'                         or 'https://' protocol.
```

Log on flower:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/celery/app/trace.py", line 412, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/celery/app/trace.py", line 704, in __protected_call__
    return self.run(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airflow/executors/celery_executor.py", line 88, in execute_command
    _execute_in_fork(command_to_exec)
  File "/usr/local/lib/python3.7/site-packages/airflow/executors/celery_executor.py", line 99, in _execute_in_fork
    raise AirflowException('Celery command failed on host: ' + get_hostname())
airflow.exceptions.AirflowException: Celery command failed on host: 172.20.38.183
```

Log on scheduler:
```
[2021-07-09 10:00:59,824] {{scheduler_job.py:1258}} ERROR - Executor reports task instance <TaskInstance: dm-bi-edw-c360-export-brg.ext_w_brg_bi_c360_t_customer_support 2021-07-08 10:00:00+00:00 [queued]> finished (failed) although the task says its queued. (Info: None) Was the task killed externally?
[2021-07-09 10:03:59,831] {{timeout.py:36}} ERROR - Process timed out, PID: 15223
[2021-07-09 10:09:11,547] {{dagrun.py:429}} ERROR - Marking run <DagRun dm-bi-edw-c360-export-brg @ 2021-07-08 10:00:00+00:00: scheduled__2021-07-08T10:00:00+00:00, externally triggered: False> failed
```

But there's no log on the worker. It seems to me like the task was never executed on the worker at all.

**Comment by ashb** [comment]
@vdusek can you look in the output/stdout/stderr of your `airflow celery worker` process? That might show the real error

**Comment by vdusek** [comment]
```
[2021-07-09 10:00:54,945: ERROR/ForkPoolWorker-32] Failed to execute task dag_id could not be found: dm-bi-edw-c360-export-brg. Either the dag did not exist or it failed to parse..
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airflow/executors/celery_executor.py", line 117, in _execute_in_fork
    args.func(args)
  File "/usr/local/lib/python3.7/site-packages/airflow/cli/cli_parser.py", line 48, in command
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airflow/utils/cli.py", line 91, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airflow/cli/commands/task_command.py", line 220, in task_run
    dag = get_dag(args.subdir, args.dag_id)
  File "/usr/local/lib/python3.7/site-packages/airflow/utils/cli.py", line 193, in get_dag
    'parse.'.format(dag_id)
airflow.exceptions.AirflowException: dag_id could not be found: dm-bi-edw-c360-export-brg. Either the dag did not exist or it failed to parse.
[2021-07-09 10:00:54,979: ERROR/ForkPoolWorker-32] Task airflow.executors.celery_executor.execute_command[22a8df15-e32b-4719-a681-7c65fec9cb5e] raised unexpected: AirflowException('Celery command failed on host: 172.20.38.183')
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/celery/app/trace.py", line 412, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/celery/app/trace.py", line 704, in __protected_call__
    return self.run(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/airflow/executors/celery_executor.py", line 88, in execute_command
    _execute_in_fork(command_to_exec)
  File "/usr/local/lib/python3.7/site-packages/airflow/executors/celery_executor.py", line 99, in _execute_in_fork
    raise AirflowException('Celery command failed on host: ' + get_hostname())
airflow.exceptions.AirflowException: Celery command failed on host: 172.20.38.183
```

Which is interesting. I checked it and DAGs are (were) there and there's no problem with their parsing.

**Comment by anirudhbagri** [comment]
@ashb, You are correct. The log file actually doesn't exist. I am also facing the same issue.
How can we fix this?

**Comment by YonatanKiron** [comment]
Same for us, exactly the same error as @vdusek posted

**Comment by xuemengran** [comment]
Hi,
I am still seeing the issue in 2.1.1 version, my executor is celery

**Comment by xuemengran** [comment]
In the 2.1.1 version, I tried to modify the airflow/utils/log/file_task_handler.py file to obtain the hostname information by reading the log table. 
I confirmed through debug that I could get the host information in this way, but a bigger problem appeared. 
The task is marked as successful without scheduling, and the log is still not viewable, so I confirm that to solve this problem, the host information must be written to the task_instance table before the task is executed. 
I think this bug is very Important, because it directly affects the use of airflow in distributed scenarios, please solve it as soon as possible!!!

**Comment by gowdra01** [comment]
I am stuck with this too I am using 2.1.0 version and getting the error
*** Failed to fetch log file from worker. 503 Server Error: Service Unavailable for url:

**Comment by knutole** [comment]
We are also getting `No SecretsMasker found!` on Airflow 2.1.2... 

Could this be due to breaking changes in the configuration file? 

We have tried setting `hide_sensitive_var_conn_fields = False` to no avail. 

``` 
[2021-08-18 11:05:53,690] {celery_executor.py:120} ERROR - Failed to execute task No SecretsMasker found!.
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/airflow/executors/celery_executor.py", line 117, in _execute_in_fork
    args.func(args)
  File "/usr/local/lib/python3.6/dist-packages/airflow/cli/cli_parser.py", line 48, in command
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/airflow/utils/cli.py", line 91, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/airflow/cli/commands/task_command.py", line 212, in task_run
    settings.configure_orm(disable_connection_pool=True)
  File "/usr/local/lib/python3.6/dist-packages/airflow/settings.py", line 224, in configure_orm
    mask_secret(engine.url.password)
  File "/usr/local/lib/python3.6/dist-packages/airflow/utils/log/secrets_masker.py", line 91, in mask_secret
    _secrets_masker().add_mask(secret, name)
  File "/usr/local/lib/python3.6/dist-packages/airflow/utils/log/secrets_masker.py", line 105, in _secrets_masker
    raise RuntimeError("No SecretsMasker found!")
RuntimeError: No SecretsMasker found!
[2021-08-18 11:05:53,710: ERROR/ForkPoolWorker-3] Task airflow.executors.celery_executor.execute_command[f6a9b0cd-bb0c-414a-a51c-80579f2d2f1e] raised unexpected: AirflowException('Celery command failed on host: 64c3bc97f173',)
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/celery/app/trace.py", line 412, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/celery/app/trace.py", line 704, in __protected_call__
    return self.run(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/airflow/executors/celery_executor.py", line 88, in execute_command
    _execute_in_fork(command_to_exec)
  File "/usr/local/lib/python3.6/dist-packages/airflow/executors/celery_executor.py", line 99, in _execute_in_fork
    raise AirflowException('Celery command failed on host: ' + get_hostname())
airflow.exceptions.AirflowException: Celery command failed on host: 64c3bc97f173
```

**Comment by potiuk** [comment]
You likely copied "verbatim" old configuration from Airlfow and modified it, so the new configuration from DEFAULT_LOGGING_CONFIG is not used.

**Comment by knutole** [comment]
Thanks @potiuk, that is exactly what happened.

**Comment by zhengxianh** [comment]
Hi, I'm trying to set up Airflow 2.1.4 with docker on multiple machines, but failed when the webserver is not able to access the task logs. Can anyone help me out please?

I've created a post in stackoverflow with details:
https://stackoverflow.com/questions/68694805/airflow-webserver-not-able-to-access-remote-worker-logs

If you need any other info, please let me know.

Thank you!

**Comment by danilocurvelo** [comment]
Exactly the same error as @vdusek posted in Airflow 2.1.x. Was this fixed in Airflow 2.2.x?

**Comment by blotouta2** [comment]
I have also got the same exception, was this issue fixed in any version ?

**Comment by potiuk** [comment]
Well. @blotouta2. I think this issue discusses 3 or 4 different. Problems. The issue is clsoed so it's likely it's a different issue altogether.

If you REALLLY want to get help, just open a new issue and provide all the details you can (and ideally a reproducible case). Or if you do not have reproducible case  - provide as much information as you can and open a Github Discussion

And BTW. if you are using an older version of Airflow, just upgrade to the newest and check - Airlfow follows SemVer, so it should be rather safe to upgrade.

**Comment by derkuci** [comment]
> In the 2.1.1 version, I tried to modify the airflow/utils/log/file_task_handler.py file to obtain the hostname information by reading the log table. I confirmed through debug that I could get the host information in this way, 

@xuemengran could you kindly point to how this could be done?

With airflow 2.2.2 + Celery, I am seeing error messages like below due to `TaskInstance.hostname` being always the latest and not relying on the `try_number`.
```
"Failed to fetch log file from worker. Client error '404 NOT FOUND' for url ..."
```
If we try really hard, the logs can be found from the local storage of _some_ celery workers.  But that is a huge burden for operational and/or debugging.

**Comment by derkuci** [comment]
@xuemengran nvm.  I guess you meant the PR suggested here.  I tried that; the log table has changed, and I couldn't match the information easily.  The `execution_date` column is None when `event="cli_task_run"`, which makes filtering impossible.

I understand why the PR was rejected.  For cases where the logs exist but the web UI couldn't locate the correct hostname, the issue is that the "task_instance" table only stores the latest `try_number`/`hostname` for a task run (as already indicated by @ITriangle).  The PK doesn't include `try_number`.  It's better to fix the task_instance table, which is more fundamental, and probably would intimidate most "amateurs" (like me).

**Pull request closed without merging**